### ISSUE TITLE
Compatibility with existing color forumulas

### ISSUE THREAD

**Issue by perrygeo** [mapbox/rio-color#4]
The color formulas for each source are defined in a `color` file containing the command line arguments to imagemagick `convert`

We need to translate these to rio color arguments. Two main options

1. ~~Have a `rio color --compatability` option that will read convert args and interpret them on the fly~~
2. Script/manually adjust all the color formulas in pxm-sources. Create an alternate `color.rio` file that mirrors the intent of `color` but with `rio color` syntax

Questions:
* what do we do with convert options that don't translate well to rio color?
* how are we going to visually test to make sure rio color output matches convert's?

**Comment by dnomadb** [comment]
Another possible tweak that I am (a) not sure how will fit in, and (b) not sure if is important, but: @camillacaros and I have talked about `jsonifying` pxm-sources` such that each source is a `json` instead of text / dirs - this would ostensibly include the color formula.

Just throwing this in there as food for 💭

**Comment by perrygeo** [comment]
Might be a nice touch to just have a little script that takes imagemagick > rio color syntax

**Comment by perrygeo** [comment]
@dnomadb re: jsonifying pxm-sources ... is that happening or on hold?

quick poll re: `rio color` and our existing imagemagickckck formulas. I'm debating three options:
* covert existing `color` files to use `rio color` options
* have `rio color` take an optional `--magick` option that would internally translate a limited subset of image magick options
* have some intermediate tool, ie `rio color $(cat color | magick2rio)`

Regardless of how we implement it, it seems like we should have a python function that does the translation. So I'll think about ☝️ and make that function first.

**Comment by dnomadb** [comment]
@perrygeo not active, but not for any sake other than there was not a strong impetus for converting it. TBH the conversion of sources to `json` itself would be v simple, I am more worried about use in pxm itself (this will be a bigger lift). That said, I could grab this soon if it helps to unblock `rio color`

**Comment by perrygeo** [comment]
@dnomadb not a blocker at all - doesn't really matter if its in a file or a json string - at some point the `convert` args gotta get translated to `rio color` args.

I made a `magick_to_color` function in the `its-magick` branch so we can do stuff like:

```
>>> from rio_color.utils import magick_to_rio
>>> magick_to_rio("-channel B -gamma 1.17 -channel R -gamma 0.96 -channel RGB -modulate 100,150 -sigmoidal-contrast 10x52%")
('gamma B 1.17', 'gamma R 0.96', 'saturation 150', 'sigmoidal R,G,B 10 0.52')

**Comment by perrygeo** [comment]
I've automated the conversion of source color files using the `magick_to_rio` function. Now the question is how well the results compare visually...

**Comment by perrygeo** [comment]
First test on the chilliwack ortho

### Original
![orig](https://cloud.githubusercontent.com/assets/1151287/14641100/a29c014c-0612-11e6-9d31-d048fbfd9018.png)

### Imagemagick convert
```
convert 13-1326-2805-test-2015-2012_30cm_594_5452.tif \
    -sigmoidal-contrast 5,30% -channel G -gamma 0.9 -channel RGB -modulate 100,125 magick.tif
```
![magick](https://cloud.githubusercontent.com/assets/1151287/14641123/bb9606c0-0612-11e6-925d-d688348749cd.png)

### Rio color
```
rio color 13-1326-2805-test-2015-2012_30cm_594_5452.tif rio.tif \
    "sigmoidal R,G,B 5 0.3" "gamma G 0.9" "saturation 125"
```
![rio](https://cloud.githubusercontent.com/assets/1151287/14641133/cbea1458-0612-11e6-8507-80b04fc7eda0.png)

**Comment by perrygeo** [comment]
Another example with a more extreme color correction and some more contrasting features

```
rio color orig.tif rio.tif \
    "gamma G 0.9" "gamma B 0.88" "gamma R 1.025" \
    "sigmoidal R,G,B 9 0.42" "saturation 125"

convert orig.tif \
    -channel G -gamma 0.9 -channel B -gamma 0.88 -channel R -gamma 1.025 \
    -channel RGB -sigmoidal-contrast 9,42% -modulate 100,125 magick.tif
```

| Original | Imagemagick convert | rio color |
|------------|------------------------------|-------------|
| ![orig tif](https://cloud.githubusercontent.com/assets/1151287/14646378/3f28e246-0627-11e6-8dc4-e4cba501e7c6.png) | ![magick tif](https://cloud.githubusercontent.com/assets/1151287/14646513/b567e718-0627-11e6-89f5-eb3f9e4fc561.png) |  ![rio tif](https://cloud.githubusercontent.com/assets/1151287/14646443/7ca003ca-0627-11e6-9b7a-af3b2b08ac47.png) |

`rio color` and `convert` are not *exactly* the same (open them in a tab, and go back<->forward to see the difference) and I'm trying to understand why. 

OTOH, they are very close and the difference is hard to discern without looking hard.

**Comment by celoyd** [comment]
Those are very close, probably close enough.

I was seeing [an RMSE of about 0.5](https://github.com/mapbox/rio-color/pull/7#issuecomment-198554408), presumably from rounding and other trivial discrepancies. But I’m seeing an RMSE of 2.17 between `imagemagick` and `rio color` in that last example. The difference seems concentrated around the most saturated areas, and I wonder if this might come down partly to our fancy saturation operator.

If these are representative, I think we’re good, but I’m not ready to sign off without a little more testing.

**Comment by perrygeo** [comment]
@celoyd I bet it's due to differences in the colorspace where we apply the saturation.

[convert --modulate](http://www.imagemagick.org/script/command-line-options.php?#modulate) uses HSL but we're using LCH here.

Are there any other [skimage colorspaces](http://scikit-image.org/docs/dev/api/skimage.color.html) that might be closer to HSL? Or should we try to do the colorspace conversion manually (and drop skimage entirely?)

**Comment by celoyd** [comment]
The LCH conversion should give better results, and I don’t think the discrepancy is a problem – there’s no color formula that would have been good with IM that will be bad with RC.

**Comment by perrygeo** [comment]
@celoyd approaching this from another angle - the performance penalty of using skimage to convert to LCH colorspace is significant. Using a naive numpy implementation to convert RGB to HSL color space is roughly 75x faster and allows us to drop skimage as a dependency. 

How much better is the saturation adjustment in the LCH space? Significant enough from HSL to justify backwards incompatibility with our existing formulas, a performance penalty and a large dependency? I'm leaning *no* and would like to explore using HSL/HSV...

A few options:

1. use LCH, deal with slight change vs imagemagick and keep skimage dependency
2. use LCH, deal with slight change vs imagemagick and experiment with implementing our own LCH conversion
3. use HSL, compatible with imagemagick, fast numpy implementation is easy to implement

**Comment by ericfischer** [comment]
Sorry to jump in here @perrygeo, but I wonder if you could make the LAB conversion cheap enough to be worthwhile. If you don't worry about the nonlinearities around 0, the LAB to sRGB conversion only amounts to:

```
// LAB to XYZ
double y = l / 100.0;
double y3 = pow(y, 3.0);
double x3 = pow((a / 500.0) + y, 3.0);
double z3 = pow(y - (b / 200.0), 3.0);

// XYZ to linear sRGB
r = (x3 *  3.2406) + (y3 * -1.5372) + (z3 * -0.4986);
g = (x3 * -0.9689) + (y3 *  1.8758) + (z3 *  0.0415);
b = (x3 *  0.0557) + (y3 * -0.2040) + (z3 *  1.0570);

// Linear sRGB to sRGB
r = pow(r, 1.0 / 2.4);
g = pow(g, 1.0 / 2.4);
b = pow(b, 1.0 / 2.4);
```

and you could even get away with skipping the exponentiation on the LAB side if you can do your saturation adjustments exponentially instead of linearly. It seems like that would be nicer than having to worry about the saturation adjustments in HSL also changing the perceived lightness as a side effect.

**Comment by perrygeo** [comment]
Thanks @ericfischer - From what I'm hearing, LCH colorspace is really the way to go in terms of quality of saturation adjustment, right? In that case, the question becomes how to do it efficiently on a large 3D array. I'll have to implement your code in numpy and see if I can make it faster than out current implementation which uses scikit-image to go `rgb -> lab -> lch`, then adjust saturation (act. chroma) linearly, then `lch -> lab -> rgb`.

**Comment by ericfischer** [comment]
Yes, I think LCH is the colorspace that will let you best adjust each of the parameters independent of the others. It should be possible to do it in xyY space too, where hue and saturation are angle and distance on the xy plane independent of lightness on Y, but LCH has the extra convenience of putting gray at the origin instead of at (⅓,⅓) and giving you perceptually uniform hue angles (in exchange for the inconvenience of not being able to mix colors by linear interpolation on the plane).

**Comment by perrygeo** [comment]
@celoyd So, wrapping this ticket up...

1. We like LCH
2. We're OK with slightly differences when applying saturation from our existing imagemagick color formulas
3. We're OK with the performance hit of converting colorspaces (at least until we can optimize it, see #21 )
4. We'll handle the details of color formula conversion outside of this repo but leverage the `magick_to_rio` function to do so.

agreed?
